spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableStrategy.scala

```diff
@@ -0,0 +1,267 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.commands
+
+import org.apache.spark.sql.delta.skipping.clustering.{ClusteredTableUtils, ClusteringColumnInfo, ClusteringStatsCollector}
+import org.apache.spark.sql.delta.skipping.clustering.ZCube
+import org.apache.spark.sql.delta.{DeltaConfigs, DeltaErrors, OptimisticTransaction, Snapshot}
+import org.apache.spark.sql.delta.actions.{AddFile, DeletionVectorDescriptor, FileAction, RemoveFile}
+import org.apache.spark.sql.delta.commands.OptimizeTableStrategy.DummyBinInfo
+import org.apache.spark.sql.delta.commands.optimize.{AddFileWithNumRecords, DeletionVectorStats, OptimizeStats, ZOrderFileStats, ZOrderStats}
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+import org.apache.spark.sql.delta.sources.DeltaSQLConf.{DELTA_OPTIMIZE_CLUSTERING_MIN_CUBE_SIZE, DELTA_OPTIMIZE_CLUSTERING_TARGET_CUBE_SIZE}
+import org.apache.spark.sql.delta.zorder.ZCubeInfo
+import org.apache.spark.sql.SparkSession
+
+object OptimizeTableMode extends Enumeration {
+  type OptimizeTableMode = Value
+  val COMPACTION, ZORDER, CLUSTERING = Value
+}
+
+/**
+ * Defines set of utilities used in OptimizeTableCommand. The behavior of these utilities will
+ * change based on the [[OptimizeTableMode]]: COMPACTION, ZORDER and CLUSTERING.
+ */
+trait OptimizeTableStrategy {
+  def sparkSession: SparkSession
+
+  /**
+   * Utility method to get max bin size in bytes to group files into.
+   */
+  def maxBinSize: Long
+
+  /**
+   * Utility method to prepare files in a partition for optimization.
+   *
+   * By default it sorts files on the size for the binpack.
+   *
+   * @return Prepared files for the subsequent optimization.
+   */
+  def prepareFilesPerPartition(inputFiles: Seq[AddFile]): Seq[AddFile] = inputFiles.sortBy(_.size)
+
+  def optimizeTableMode: OptimizeTableMode.Value
+
+  /**
+   * The clustering algorithm to be used by either by ZORDER or Liquid CLUSTERING.
+   * An error is thrown for COMPACTION.
+   */
+  def curve: String
+
+  /**
+   * Prepare a new Bin and returns its initialized [[BinInfo]].
+   */
+  def initNewBin: OptimizeTableStrategy.BinInfo = DummyBinInfo()
+
+  /**
+   * Incorporate essential tags for optimized files based on the [[OptimizeTableMode]].
+   */
+  def tagAddFile(file: AddFile, binInfo: OptimizeTableStrategy.BinInfo): AddFile =
+    file.copy(dataChange = false)
+
+  /**
+   * Utility to update additional metrics after optimization.
+   *
+   * @param optimizeStats The input stats to update on.
+   * @param removedFiles Removed files.
+   * @param bins Bin-packed file groups from [[groupFilesIntoBins]].
```

Comment:
it's weird to reference a specific caller from the interface. can you just describe what's here (e.g., what are the key/value pairs, etc.)